File: src/memory.ts
export class Memory {
    private ramBuffer = new ArrayBuffer(4 * 1024 * 1024); // 4MiB RAM
    private ram = new DataView(this.ramBuffer);
    private rom: DataView;

    private romMapped = true;

    constructor(romData: ArrayBuffer) {
        this.rom = new DataView(romData);
        console.debug('ROM size: 0x' + this.rom.byteLength.toString(16));
    }

    load(address: number): number {
        address &= 0x03fffffc;

        // Logically mapped RAM unless ROM is mapped low
        if (address < 0x02000000) {
            if (this.romMapped) {
                console.debug("Fetching from ROM mapped low");
                return this.rom.getUint32(address, true);
            }

            // This should be mapped logical-to-physical
            console.debug("Fetching from logical RAM");
            return this.ram.getUint32(address, true);
        }

        // Physically mapped RAM
        if (address < 0x03000000) {
            console.debug("Fetching from physical RAM");
            return this.ram.getUint32(address - 0x02000000, true);
        }

        // I/O controllers
        if (address < 0x03400000) {
            console.debug("Fetching from I/O controllers");
            return 0;
        }

        // Low ROM
        if (address < 0x03800000) {
            console.debug("Fetching from low ROM");
            return 0;
        }

        // High ROM
        let romAddress = address - 0x03800000;
        console.debug(`Fetching from high ROM ${romAddress.toString(16)}`);

        if (romAddress < this.rom.byteLength) {
            return this.rom.getUint32(romAddress, true);
        } else {
            console.warn("Read past end of ROM; returning 0");
            return 0;
        }
    }

    store(address: number, data: number): void {
        address &= 0x03fffffc;
        this.ram.setUint32(address, data, true);
    }

    loadByte(address: number): number {
        let word = this.load(address);
        let field = address & 3;
        return (word >> (field * 8)) & 0xff;
    }

    storeByte(address: number, data: number): void {
        let field = address & 3;
        let word = (data & 0xff) << (field * 8);
        this.store(address, word);
    }
}

File: src/webarc/memory.rs
use std::ops::Shl;
use std::ops::Shr;

pub struct Memory {
    ram: Box<[u32]>,
    rom: Box<[u32]>,
    rom_mapped: bool
}

impl Memory {
    pub fn new(rom: Box<[u32]>) -> Memory {
        Memory {
            ram: vec![0; 1 * 1024 * 1024].into_boxed_slice(),
            rom,
            rom_mapped: true
        }
        // console.debug('ROM size: 0x' + this.rom.byteLength.toString(16));
    }

    pub fn load(&mut self, address: u32) -> u32 {
        let masked_address = address & 0x03fffffc;

        // Logically mapped RAM unless ROM is mapped low
        if masked_address < 0x02000000 {
            if self.rom_mapped {
                // console.debug("Fetching from ROM mapped low");
                self.rom[(masked_address / 4) as usize]
            } else {
                // This should be mapped logical-to-physical
                // console.debug("Fetching from logical RAM");
                self.ram[(masked_address / 4) as usize]
            }
        } else if masked_address < 0x03000000 {
            // Physically mapped RAM
            // console.debug("Fetching from physical RAM");
            self.ram[((masked_address - 0x02000000) / 4) as usize]
        } else if masked_address < 0x03400000 {
            // console.debug("Fetching from I/O controllers");
            unimplemented!("Reading from I/O controllers")
        } else if masked_address < 0x03800000 {
            // console.debug("Fetching from low ROM");
            self.rom_mapped = false;
            unimplemented!("Low ROM")
        } else {
            // High ROM
            self.rom_mapped = false;
            self.rom[((masked_address - 0x03800000) / 4) as usize]
        }
    }

    pub fn store(&mut self, address: u32, _data: u32) {
        let masked_address = address & 0x03fffffc;
        println!("Store address: {:08X}", masked_address);

        // Logically mapped RAM unless ROM is mapped low
        if masked_address < 0x02000000 {
            unimplemented!("Writing to logically mapped RAM");
        } else if masked_address < 0x03000000 {
            unimplemented!("Writing to physically mapped RAM");
        } else if masked_address < 0x03400000 {
            // console.debug("Fetching from I/O controllers");
            unimplemented!("Writing to I/O controllers")
        } else if masked_address < 0x03600000 {
            unimplemented!("Writing to VIDC");
        } else if masked_address < 0x03800000 {
            self.rom_mapped = false;
            unimplemented!("Writing to DMA/MEMC");
        } else {
            self.rom_mapped = false;
            unimplemented!("Writing to L2P address translator");
        }
    }

    pub fn load_byte(&mut self, address: u32) -> u8 {
        let word = self.load(address);
        let field = address & 0x00000003;
        (word.shr(field * 8) & 0xff) as u8
    }

    pub fn store_byte(&mut self, address: u32, data: u8) {
        let field = (address & 0x00000003) as u8;
        let word = (data & 0xff).shl(field * 8);
        self.store(address, word as u32);
    }
}
